http/org.wso2.carbon.transport.http.netty/src/main/java/org/wso2/carbon/transport/http/netty/sender/channel/pool/ConnectionManager.java

```diff
@@ -232,7 +232,10 @@ private void releaseChannelToPool(TargetChannel targetChannel, GenericObjectPool
     public void invalidateTargetChannel(TargetChannel targetChannel) throws Exception {
         Map<String, GenericObjectPool> objectPoolMap = targetChannel.getCorrelatedSource().getTargetChannelPool();
         try {
-            objectPoolMap.get(targetChannel.getHttpRoute().toString()).invalidateObject(targetChannel);
+            // Need a null check because SourceHandler side could timeout before TargetHandler side.
+            if (objectPoolMap.get(targetChannel.getHttpRoute().toString()) != null) {
```

Comment:
Nope. Because SourceHanlder and the respective TargetHandler runs on the same thread.